### PR TITLE
[SPARK-10671] [SQL] Throws an analysis exception if we cannot find Hive UDFs

### PR DESCRIPTION
**Pull request by cloud-fan**
Takes over https://github.com/apache/spark/pull/8800

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yhuai

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveUDFSuite.scala [hunk 1]
```diff
@@ -299,6 +300,42 @@ class HiveUDFSuite extends QueryTest with TestHiveSingleton {
...
+    }
+
+    {
+      // Hive UDTF
+      sql(s"CREATE TEMPORARY FUNCTION testUDTFExplode AS '${classOf[GenericUDTFExplode].getName}'")
+      val message = intercept[AnalysisException] {
+        sql("SELECT testUDTFExplode() FROM testUDF")
+      }.getMessage
+      assert(message.contains("No handler for Hive udf"))
+    }
```
Comment (yhuai) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveUDFSuite.scala]: Can we also add tests for `GenericUDF` and `AbstractGenericUDAFResolver` (we have tests of `UDAF`, `UDF`, and `GenericUDTF` at here)?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala [hunk 2]
```diff
@@ -60,20 +60,36 @@ private[hive] class HiveFunctionRegistry(underlying: analysis.FunctionRegistry)
...
+          udtf
+        } else {
+          throw new AnalysisException(s"No handler for udf ${functionInfo.getFunctionClass}")
+        }
+      } catch {
+        case analysisException: AnalysisException =>
+          // If the exception is an AnalysisException, just throw it.
+          throw analysisException
+        case throwable: Throwable =>
+          // If there is any error, we throw an AnalysisException.
```
Comment (yhuai) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/hiveUDFs.scala]: If there is any error => If there is any other error

**Comment by yhuai** [comment]
LGTM. Merging to master.

**Pull request closed without merging**